File: parenthesesConversion.cpp
#include <string>
#include <vector>
using namespace std;

int isRightP(string u) {
    string s;
    for (int i = 0; i < u.size(); i++) {
        if (s.back() == '(' && u[i] == ')')
            s.pop_back();
        else
            s.push_back(u[i]);
    }
    if (s.empty()) return 1;
    return 0;
}

string reverse(string u) {
    u.erase(0, 1);
    u.pop_back();
    for (int i = 0; i < u.size(); i++) {
        if (u[i] == '(')
            u[i] = ')';
        else
            u[i] = '(';
    }
    return u;
}

pair<string, string> getP(string p) {
    string u = "";
    string v = "";
    int cnt = 0, i = 0;
    for (; i < p.size(); i++) {
        if (i != 0 && cnt == 0) break;
        u += p[i];
        if (p[i] == '(')
            cnt++;
        else
            cnt--;
    }
    for (; i < p.size(); i++) v += p[i];
    return {u, v};
}

string foo(string p) {
    if (p.size() == 0) return "";
    pair<string, string> uv = getP(p);
    if (isRightP(uv.first))
        return uv.first + foo(uv.second);
    else {
        string s = "(";
        s += foo(uv.second);
        s += ")";
        s += reverse(uv.first);
        return s;
    }
}

string solution(string p) {
    return foo(p);
}

File: parenthesesConversion.js
function isRightP(u) {
    let stack = [];
    for (let i = 0; i < u.length; i++) {
        if (stack[stack.length - 1] === "(" && u[i] == ")") stack.pop();
        else stack.push(u[i]);
    }
    if (stack.length === 0) return 1;
    else return 0;
}

function getReverse(u) {
    u.shift();
    u.pop();
    for (let i = 0; i < u.length; i++) {
        if (u[i] === "(") u[i] = ")";
        else u[i] = "(";
    }
    return u;
}

function getP(p) {
    let u = [];
    let v = [];
    let cnt = 0;
    let i = 0;
    for (; i < p.length; i++) {
        if (i !== 0 && cnt === 0) break;
        u.push(p[i]);
        if (p[i] === "(") cnt++;
        else cnt--;
    }
    for (; i < p.length; i++) v.push(p[i]);
    return [u, v].slice();
}

function foo(p) {
    if (p.length === 0) return [];
    let [u, v] = getP(p);
    if (isRightP(u)) {
        return u.concat(foo(v));
    } else {
        let arr = ["("];
        arr = arr.concat(foo(v));
        arr.push(")");
        return arr.concat(getReverse(u));
    }
}

function solution(p) {
    return foo(p.split("")).join("");
}
